### ISSUE TITLE
Add support for results handling

### ISSUE THREAD

**Issue by mauroservienti** [ServiceComposer/ServiceComposer.AspNetCore#149]
Similar to the `IHandleResults` in the "ServiceComposer.AspNetCore.Mvc" package

**Comment by markphillips100** [comment]
Would it also be possible @mauroservienti to enable a composition handler to supply the returned http status code result?I was thinking of use cases where I might want to propagate some 400 errors (problem details) back to the caller.  At the moment I can only throw an exception which will give a 500 when I'd prefer a BadRequest 400.

**Comment by markphillips100** [comment]
I see you have something for status code handling on the [project backlog](https://github.com/orgs/ServiceComposer/projects/1#card-32184550).  Anything I can do in the meantime to obtain the desired behaviour?

I have a need to propagate a downstream API 404 result back up the call stack to the client.  At the moment the client will only ever get a 200 with empty json, or 500 if I throw exception on detecting a downstream 404.

**Comment by mauroservienti** [comment]
@markphillips100 I've been thinking about this for a while and I'm having a bit of a hard time in finding the right API.

But before diving into that. The workaround to your issue should be as easy as a middleware registered before the ServiceComposer one.

Requests handler can add a custom response header with the desired response status and the middleware after the `next()` invocation can check if the header is present and override the response status code.

**Comment by markphillips100** [comment]
Ah nice! I was modifying my client code to check for a 200 with empty content and modify its behaviour accordingly. I like your middleware approach way more than my current hack :-)

As for the "right API", perhaps a response extension to set the header, and rather than middleware, handle the header in the CompositionHandler to send the appropriate status?

**Comment by mauroservienti** [comment]
@markphillips100 I raised #168 to add support for custom status codes. In your handlers, you can now do something like:

```
var response = request.HttpContext.Response;
response.StatusCode = (int)HttpStatusCode.Forbidden;
```

and the custom status code won't be overwritten anymore. The middleware trick doesn't work, unfortunately, as by the time the middleware is invoked the response has already been started and the status code cannot be changed anymore.

**Comment by markphillips100** [comment]
Thanks for this Mauro.  I'll give it a go ASAP.

As for the middleware approach, I did get an implementation working, not that it matters now ;-).  Did you use the OnStarting response delegate in the middleware?

```
    public class ComposedStatusCodeHeaderMiddleware
    {
        private readonly RequestDelegate _next;

        public ComposedStatusCodeHeaderMiddleware(RequestDelegate next)
        {
            _next = next ?? throw new ArgumentNullException(nameof(next));
        }

        public async Task Invoke(HttpContext context)
        {
            context.Response.OnStarting(() => {
                var isSuccessStatusCode = context.Response.Headers.GetComposedIsSuccessStatusCode();
                if (!isSuccessStatusCode)
                {
                    // Composition has included an error so modify the response accordingly
                    var statusCode = context.Response.Headers.GetComposedStatusCode();

                    context.Response.StatusCode = (int)statusCode;
                }

                return Task.CompletedTask;
            });

            await _next(context);
        }
    }
```

**Comment by mauroservienti** [comment]
I was not aware of the `OnStarting` option, I learned something new!

**Comment by mauroservienti** [comment]
This can be closed as #168 fixes the problem
